### ISSUE TITLE
tf.fft2d not working properly for complex64 data

### ISSUE THREAD

**Issue by isaacgerg** [tensorflow/tensorflow#28192]
**System information**
- Windows 7
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): print(tensorflow.GIT_VERSION, tensorflow.VERSION)
b'unknown' 1.13.1
- Python version: 3.7.3 (default, Mar 27 2019, 17:13:21) [MSC v.1915 64 bit (AMD64)]
- CUDA 10, cudnn 64_10
- GPU model and memory: Titan X pascal

**Describe the current behavior**

fft2d produces non-sensical results on complex64 datatype. (also verified complex128 broken)

**Code to reproduce the issue**

```
import os
os.environ['CUDA_VISIBLE_DEVICES'] = '0'
import numpy as np
import scipy as sp
import scipy.fftpack
import tensorflow as tf
import matplotlib.pyplot as plt

#===================================================================================================
# Generate data
slc = sp.misc.face(gray=True).astype('float32')
slc = (slc + 1j*slc)[0:256, 0:256]
tileSize = slc.shape[0]

#===================================================================================================
freqImg = np.fft.fft2(slc)

#===================================================================================================
# Now do this in keras/tf
input1 = tf.keras.layers.Input(shape=(tileSize, tileSize, 1), dtype='complex128')
out = tf.keras.layers.Lambda(lambda x:tf.spectral.fft2d(x))(input1)
model = tf.keras.models.Model(input1, out)
freqImg_tf = model.predict(slc[None,:,:,None])

plt.figure(); plt.imshow(np.log(np.abs(freqImg)+1)); plt.title('sp')
plt.figure(); plt.imshow(np.log(np.abs(freqImg_tf[0,:,:,0])+1)); plt.title('keras'); plt.show()

#===================================================================================================
# Plots




print('done')
```

**Comment by isaacgerg** [comment]
based on the symmetrical output of tf.fft2d, I suspect it may be doing a real FFT instead of a complex.

**Comment by rryan** [comment]
Thanks for the report and the code example! 

`tf.signal.fft2d` operates on the inner-most 2 dimensions of the provided tensor. It looks like the tensor you are providing is rank 4, with shape `[batch, tileSize, tileSize, 1]`, so the FFT is over`[tileSize, 1]` matrices, which there are `batch * tileSize` of. This explains why you're seeing only one axis being transformed. The FFT over the length 1 axis is a no-op, so it's effectively a 1D FFT.

Here is a [Colab notebook](https://colab.research.google.com/drive/1vm3_GydusmRL5Y_7CeISunAqgLzmp0eK) with a fix to your code example (remove the inner-most 1 dimension). I realize you're probably using the layout `[batch, height, width, channels]` because that's what the TF conv ops use (`NHWC` layout), and I'm sorry for the confusion, but the FFT ops don't assume that layout. We follow `numpy.fft` convention and operate on the inner-most N axes for an N-D FFT (and unfortunately we don't support an axis argument for FFTs yet).

**Comment by isaacgerg** [comment]
Hi @rryan, the documentation states the fft2 operates on the innermost dimensions, so wouldn't that be [tileSize, tileSize]?  I can only think that our definition of "innermost" is different.  

Thanks for the fix and explanation.  Perhaps the documentation should be updated?  (i found it hard to believe when I googled for this answer that no one else had this issue.)  I didnt think to cross check it with numpy.fft -- good idea.  In any case, I'll update my code tomorrow and ensure the fix works.  Thank you!

**Comment by rryan** [comment]
@isaacgerg it sounds like we have a terminology difference then! TensorFlow shapes are generally described as `[outermost, outer, middle, inner, innermost]`, so innermost 2 dimensions refers to `[inner, innnermost]` for that example. Woud "last" instead of "inner" make more sense to you?

**Comment by isaacgerg** [comment]
@rryan  Correct.
